File: Coin Change/coin change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        return f(amount, coins, dp);
    }
    
    int f(int i, int[] coins, int[] dp) {
        if (i <= 0) {
            return 0;
        }
        else {
            int ans = -1;
            if (dp[i] != 0) {
                return dp[i];
            }
            for(int coin: coins) {
                if (coin <= i) {
                    int val = f(i - coin, coins, dp);
                    if (val != -1 && (ans == -1 || val + 1 < ans)) {
                        ans = val + 1;
                    }
                }
            }
            dp[i] = ans;
            return ans;
        }
    }
}

File: Coin Change/coin change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        self.dp = { }
        self.coins = coins
        return self.f(amount)
        
    def f(self, i):
        if (i <= 0):
            return 0
        else:
            ans = -1
            if (i in self.dp):
                return self.dp[i]
            else:
                for coin in self.coins:
                    if (coin <= i):
                        val = self.f(i - coin)
                        if (val != -1 and (ans == -1 or val + 1 < ans)):
                            ans = val + 1
                self.dp[i] = ans
                return ans